### ISSUE TITLE
Flow does not accept intersection function types even with disjoint unions

### ISSUE THREAD

**Issue by danwang** [facebook/flow#3021]
Hi!

I have the following code ([try link](https://flowtype.org/try/#0C4TwDgpgBAglC8UDeAoKVSQFxQOQENcAaNKfHAZ2ACcBLAOwHMSBfAbhRU2gCEFlS3HLgBGxUiJz0ArgFsREaqw5dw0AGL8AFFvKwAlAgB8UKnSaGAZFB2SoPQ-BMz5i-RwDGAe3pUoAMy8vHE1EXTs4AB97RxNUdFp-G3wRADpuBHhEAlxDePQoaghgaWp6MjT8DnQWKAgAGwpofPQikrKK1JFqqBYUFiA))

```jsx
type A = {
  type: 'a',
  a: string,
};

type B = {
  type: 'b',
  b: number,
};

type F = ((a: A) => string) & ((b: B) => number);
const foo: F = (ab: A | B) => {
  if (ab.type === 'a') {
    return ab.a;
  } else {
    return ab.b;
  }
}
```

I wish to make a function `foo` accepting a disjoint union but which returns different types for each member.

This fails with the following error:
```
14:     return ab.a;
               ^ string. This type is incompatible with
11: type F = ((a: A) => string) & ((b: B) => number);
                                             ^ number
16:     return ab.b;
               ^ number. This type is incompatible with
11: type F = ((a: A) => string) & ((b: B) => number);
                        ^ string
```

What's the best way to express this type of function (disjoint types input, but different output type for each one)? I would like to not have to declare it as `(ab: A|B) => string|number` because that loses some information.

Sincerely,
Dan

**Comment by danwang** [comment]
bump

**Comment by samwgoldman** [comment]
I don't think this is a bug.

The program in the original post asks whether the following subtyping rule is true:

```
(A|B => X|Y) <: ((A => X) & (B => Y))
```

The above is not true, but the reverse is true.

**Comment by rhendric** [comment]
I agree that the subtyping relation you wrote is not true, but I think the spirit of the bug is that there's no way to write `foo` so that Flow infers it to have the desired type `F`, explicitly or implicitly (without cheating). Is that not the case? I would be overjoyed if not.

**Comment by jcready** [comment]
Yeah, the feature request here is to get flow to accept a real javascript function as an intersection of function types. You can _declare_ function intersections and you can cheat by first casting a function to `any` and then to a function intersection, but then you lose all guarantees that the function actually adheres to the intersection constraints. Flow should be able to verify that a real javascript function adheres to a function intersection type.

**Comment by zeorin** [comment]
I noticed something interesting recently:

```js
// @flow

interface T {
	(t: number): number;
	(t: string): string;
}

declare var returnTypeInterface: T;

(returnTypeInterface('string'): string);
(returnTypeInterface(2): number);

// $ExpectError
(returnTypeInterface('string'): number);
// $ExpectError
(returnTypeInterface(2): string);
```
<sup>([Flow Try](https://flow.org/try/#0PTAEAEDMBsHsHcBQiCWA7ALgUwE6QIYDGWoAKqAN6ICQAFBgFyhoCuAtgEa4CUTrnuANw16TAM4Yc6AOa9QEqWmnCAvsgAmWQtHw4SAN12g9GFjjSkAngAcsASUy4CxJqWGJaJsxZv3HeIixaAHIFGWC5MKVuYU8sU3MrWwdsAOJaACY5fi4cGOQQUAASAFEAD1tCDBKcHFgcDy9E3xSnQJCo6Qi+dlz8wtKKrWra+sb47yS-VOcgrPFJGRigA))</sup>

This seems to work. But I'm not sure if it's safe.

**Comment by apsavin** [comment]
@samwgoldman @mrkev can we reopen this issue?

**Comment by DeedleFake** [comment]
I think an issue I'm having is related to this, so I'm asking about it here:

I'm trying to get a function to return different, but related, types based on the type of the input. For example:

```javascript
type Arg = { +[string]: Object }
type CallbackArg = () => Arg
type Func = <T: Arg | CallbackArg>(T) => /* Here's where I'm stuck. */
```

What I'd like to do is return `{ +[$Keys<T>]: string }` if the object is passed directly and `{ +[$Keys<$Call<T>>]: string}` if a function is passed. Is there any way to do this type of conditional overloading? I haven't been able to find anything obvious via either the documentation or Google. I'm essentially looking for some kind of `$IfCompatible<T, Arg, { +[$Keys<T>]: string }, { +[$Keys<$Call<T>>]: string }>`, but a feature like that doesn't seem to exist. Am I going about this completely wrong?

For clarification, by the way, I'm dealing with this inside a type declaration file, so I can't write some kind of utility function to handle the conditional for me via `if`s and a `$Call<>`.

I was able to get it _mostly_ working via

```javascript
type Func = <T: Arg | CallbackArg>(T) => { +[$Keys<T> | $Keys<$Call<T>>]: string }
```

but it seems kind of awkward. I'm also not sure why this would work, as I would think Flow would complain about `$Call<T>` making no sense if `T` is `Arg`. For the record, using `$Keys<T | $Call<T>` does _not_ work.

([Try](https://flow.org/try/#0PTAEAEDMBsHsHdQGcAuAnAlgYxQKFygJ4AOApqAIJoDmoAvKAN6gDUA2qpgHbUC6AXKADyAIwBWpHKAC+BEuSrUAwgENo0ESqwBreqAAUASnoA+SjTllQAMQCuXLHoA8AFUGLQAH3PK1Grdom+i7GdGbM7AAkANKkhEiuZt4xcQmRquqJJgLI6Bg8MvhYsFyooJCCdg56+iih4bigTdCkKKBoenhNoBiQBkRksH0ddKOgAOSQ9jgYJePGjI3d7XpoRkuF3RtorbZoXMLikigAdNqp+miGJzsAJrZYpPr6WlgANKBYe-UGjCf-rw+bC+aBy43G0kMH0YkNwsmWuBabTaDEgzx++kYKkEAEYPiJBAAmSGGAgnFS4ECgSIAUQAHmQcDS0GhYGgyQBHfAo8qYjm4j7wImw04Uqm0hnHZms9mnDlAA))

**Comment by rhendric** [comment]
An interface is probably your best bet: [Try](https://flow.org/try/#0PTAEAEDMBsHsHdQGcAuAnAlgYxQKFygJ4AOApqAIJoDmoAvKAN6gDUA2qpgHbUC6AXKADyAIwBWpHKAC+uDFxSk0kAIZZyAMQCuXLE1yhQAHgAqgqtQB8ACmsBKepdAm7g5uwAkAaVKEkpywFkdHlaaQBuA2MzShobFzdWNm9ffxNAwU5QmUjZfCxYLlRQSEFtXXpQaxQHOidGKNBoUhRQNEq8Q1AMSCqiMlhe9roR0ABySB0cDEKxhwauw2G2+0bZLsa0Fq00LmFxSRQAOgBrVOs0OyOtgBMtdVs1LAAaUCwd2qdrRiPfp9e2O80EExmNpHZXoxwbh1htmq1WgxILZPlVGCpBABGV4iQQAJnBdgIRxUuBAoA8AFEAB5kHCUtBoWBoMnAY4AR3wiJK33ZWNe8Hx0JAx1J5KptMODKZLI5QA)

**Comment by DeedleFake** [comment]
Huh. That should probably be in the documentation.

Oddly, [the order seems to affect it](https://flow.org/try/#0PTAEAEDMBsHsHdQGcAuAnAlgYxQKFygJ4AOApqAIJoDmoAvKAN6gDUA2qpgHbUC6AXKADyAIwBWpHKAC+uDFxSk0kAIZZyAMQCuXLE1yhQAHgAqgqtQB8AChMBKQc3YASANKlCSU5YHJ082mkAbgNjM0oaG2s7ektQe0dWNjcPLxMfQU4AmRDZXCxYLlRQSEFtXXpQaxQYujjGUNBoUhRQNEq8Q1AMSCqiMlhe9roR0ABySB0cDEKxmIauw2G26MbZLsa0Fq00LmFxSRQAOgBrVOs0OyOtgBMtdWtrNSwAGlAsHdq460Yjv+e3mwPmhfGMxtI7G9GBDcOsNs1Wq0GJBHl8qowVIIAIxvESCABMELsBCOKlwIFAzgAogAPMg4KloNCwNAkgCO+CRJR+bOxb3gBJhxzJFOpdMOjOZrOObKAA). If the variant that takes `() => T` is listed in the interface definition second, it'll fail again.

**Comment by FireyFly** [comment]
Interesting question! An interface is probably cleanest, I didn't realise they work that way either... good to know. I *think* it essentially desugars to an intersection type; [this (try)][1] is what I came up with when experimenting.

The relevant bit being:
```js
type ObjectToString = Object => string;
type Func =
  & (<T: Arg>(() => T) => $ObjMap<T, ObjectToString>)
  & (<T: Arg>(T) => $ObjMap<T, ObjectToString>);
```

@DeedleFake wrt order, I believe they're tried left-to-right (or top-down in the case of multiple declarations in an interface I suppose), so you probably want to put more specific types first.

[1]: https://flow.org/try/#0PTAEAEDMBsHsHdQGcAuAnAlgYxQKFygJ4AOApqAIJoDmoAvKAN6gDUA2qpgHbUC6AXKADyAIwBWpHKAC+AbgIlyVagGEAhtGgi1WANb1QACgCU9AHyUa8hWWHjJKACqwAyugw8DoiVLoXOHtTyRLYAYgCuXFj0uKCgAGRGADyOgspmhibmoI6mfqAAJN4AsmrEKQA0dj5Oru48ZsaxCcmpltQZudlF4qXljlXeDs5u3B3G1liwXKigkIIRUQaGKHkWjM2gUzMooGhpNAYhpLCQoLt0l6AA5JCROBjT16AA-OdZgijycXukKOFoLjVBwAOl0pEISEMaGMILQpAAJuEsKRMjosFUsAC1kZGCD8eiqmwsWgBDdrtJjFVGJTcNJNqBcNA-ucDJBMjjDIw1IIAIxVESCABMlKaKBBalwIEKAFEAB5kHAytBoWBoAgggCO+Aucy5mr5VXgwtp4sl0oK8sVKGVqvV4s1QA

**Comment by zeorin** [comment]
I've also been using intersection function types for this recently. I wish the docs mentioned this behaviour.

I figure an intersection object type is really used to check whether something is _more than one (inexact) object type at the same time_.

So for an intersection function type, it's really more than one function type at the same time. Of course, in my own code that's usually a good sign I'll be better off just creating individual functions instead, but sometimes I have to type other people's code, Redux + Redux Thunk's dispatch being a good example.
